### PR TITLE
Use new param log_all in CSVLogger to log all elements.

### PR DESCRIPTION
**Pull request by PiyushDatta**
@rchao Hi, can you please review the following. Reference to #40114 .

New feature. Use new param "log_all" in CSVLogger to log all elements in training even if some epochs don't contain the same elements. Example of this is when validation_freq is set to > 1, and some epochs won't contain validation accuracy/loss. CSV file header will contains all elements in all epochs.

### CODE REVIEW

**Comment by PiyushDatta** [comment]
@gbaned hi any update on this?

File: tensorflow/python/keras/callbacks.py [hunk 1]
```diff
@@ -2442,15 +2452,30 @@ class CustomDialect(csv.excel):
...
-    self.csv_file.flush()
+    row_dict.update((key, handle_value(logs[key]))
+                    for key in self.keys if key in logs)
+    # if user wants to log all, then we write all rows to csv file
+    # at the end of training
+    if not self.log_all:
+      self.writer.writerow(row_dict)
+      self.csv_file.flush()
+    else:
+      self._row_dicts.append(row_dict)
```
Comment (rchao) [inline tensorflow/python/keras/callbacks.py:2469]: Thanks for the PR! The argument log_all seems to be a behavior change: now the rows are not necessarily logged at the end of every epoch, even if the keys are present in logs. Am I understanding correctly? If so, is this expected?

**Comment by gbaned** [comment]
@PiyushDatta  Can you please check @rchao's comments and keep us posted ? Thanks!

**Comment by PiyushDatta** [comment]
Will get to this during the weekend.

File: tensorflow/python/keras/callbacks.py (hunk 1 above)
Comment (PiyushDatta) [reply tensorflow/python/keras/callbacks.py]: @rchao Thanks for the review. After looking over this, it's actually incorrect. I've put a new revision, please review that. 

The idea behind this original PR was that we want to log all keys/columns in the csv file to maintain the CSV file format (same # of columns for each row). However, with this csv callback, you only get the columns per epoch. Meaning that if we have validation freq = 3, you will have rows 1 and 2 in the csv file with maybe 3 metrics lets say, and then you'll have row 3 with the original 3 metrics + 2 additional metrics for validation epoch (val acc and val loss). Now what happens is that you have row 1 and 2 with 3 columns, and row 3 with 5 columns. This would be an invalid csv file since not all rows have the same # of columns.

To initially resolve this issue, I thought it would be okay to just keep all the data in-memory and just write everything out at the end. By the end we would know the max amount of columns, and therefore any data rows that didn't have the column, would just insert as empty string. This way we would have a valid csv file at the end.

Problem with this approach:
- Incorrectly assumed data could fit in memory
- If we ever crash before the end of training, all data is gone

**Comment by PiyushDatta** [comment]
**Validation**

```pylint --rcfile=tensorflow/tools/ci_build/pylintrc tensorflow/python/keras/callbacks.py```
Your code has been rated at 9.75/10 (previous run: 9.75/10, +0.00)

`python tensorflow/python/keras/callbacks_test.py KerasCallbacksTest.test_CSVLogger`

INFO:tensorflow:time(__main__.KerasCallbacksTest.test_CSVLogger): 1.08s
I0502 17:29:58.183570 32040 test_util.py:2102] time(__main__.KerasCallbacksTest.test_CSVLogger): 1.08s
[       OK ] KerasCallbacksTest.test_CSVLogger
Ran 1 test in 1.082s
**OK**

**Comment by PiyushDatta** [comment]
New approach:
- Write to csv file on each epoch
- If we have log_all enabled and we hit a point where the current epoch has more keys/columns than previous epochs and rows we:
       1. Create a backup file of current csv file (in case something goes wrong in this process)
       2. Create a copy of current csv file as a temp file (we will make this current csv file at end)
       3. Include new keys/columns in `self.keys`
       4. Write new columns/adjustments in to temp file and read from current csv file
       4. Go to header position in temp file (first header we wrote in this file) and write out new header with new columns
       5. Go to starting position in both temp file and csv file (first row we inserted during this training)
       6. For each row from csv file, read the row, add new columns with empty string, and write into temp file and flush
       8. Repeat for all rows until current row
       9. Close temp file, rename temp file as current file, and remove back-up file
      10. Now the temp file is renamed to current csv file, we now write the current epoch data into the file and go on with regular operations.
   
If we don't have log_all enabled or log_all is enabled but all epochs have same # of keys/columns then we keep original design/process.

**Comment by PiyushDatta** [comment]
@rchao please review new revision. @gbaned fyi.

**Comment by gbaned** [comment]
@PiyushDatta Can you please resolve conflicts? Thanks!

**Comment by PiyushDatta** [comment]
Will resolve conflicts this weekend

**Comment by mihaimaruseac** [comment]
This should be moved to the Keras repo now with the split.

CC @qlzh727

**Comment by qlzh727** [comment]
Thanks for the PR. Keras code has been moved to keras-team/keras repository, and the old code in tensorflow/python/keras will be removed soon. Please send your PR to the new repo and we will take it from there. Thanks.
